From 9fd6fc63cee60c96dcd26491390ec7312a3e25dc Mon Sep 17 00:00:00 2001
From: Diogo Ferreira <defer@cyngn.com>
Date: Mon, 15 Sep 2014 15:44:39 +0100
Subject: [PATCH] mediatek: Port AV changes

This ports the changes required to perform video decoding
and enconding.

The changes are ported from the mediatek BSP for mt6592
with the minimum required feature set and confined to
allow co-existance with changes from other vendors.

[Trimmed down for L]

[Ported to M (CM13.0) by @xen0n]

Change-Id: I3709de0e5b9e4e0f68a71e182549e72a3dab26a7

Conflicts:
	media/libstagefright/ACodec.cpp
	media/libstagefright/AwesomePlayer.cpp
	media/libstagefright/OMXCodec.cpp
	media/libstagefright/include/AwesomePlayer.h
---
 media/libstagefright/ACodec.cpp              |  6 ++++++
 media/libstagefright/AwesomePlayer.cpp       | 21 ++++++++++++++++++++-
 media/libstagefright/OMXCodec.cpp            |  9 +++++++++
 media/libstagefright/include/AwesomePlayer.h |  8 ++++++++
 4 files changed, 43 insertions(+), 1 deletion(-)

diff --git a/media/libstagefright/ACodec.cpp b/media/libstagefright/ACodec.cpp
index 427728f08d4..7f01b8cda4a 100644
--- a/media/libstagefright/ACodec.cpp
+++ b/media/libstagefright/ACodec.cpp
@@ -929,6 +929,7 @@ status_t ACodec::setupNativeWindowSizeFormatAndUsage(
     int32_t width = 0, height = 0;
     int32_t isAdaptivePlayback = 0;
 
+#ifndef MTK_HARDWARE
     if (mInputFormat->findInt32("adaptive-playback", &isAdaptivePlayback)
             && isAdaptivePlayback
             && mInputFormat->findInt32("max-width", &width)
@@ -940,6 +941,11 @@ status_t ACodec::setupNativeWindowSizeFormatAndUsage(
         width = def.format.video.nFrameWidth;
         height = def.format.video.nFrameHeight;
     }
+#else
+    width = def.format.video.nStride;
+    height = def.format.video.nSliceHeight;
+#endif
+
     err = setNativeWindowSizeFormatAndUsage(
             nativeWindow,
             width,
diff --git a/media/libstagefright/AwesomePlayer.cpp b/media/libstagefright/AwesomePlayer.cpp
index 933f2413b0f..de2e0a61eff 100644
--- a/media/libstagefright/AwesomePlayer.cpp
+++ b/media/libstagefright/AwesomePlayer.cpp
@@ -217,7 +217,12 @@ AwesomePlayer::AwesomePlayer()
       mTextDriver(NULL),
       mOffloadAudio(false),
       mAudioTearDown(false),
-      mIsFirstFrameAfterResume(false) {
+      mIsFirstFrameAfterResume(false)
+#ifdef MTK_HARDWARE
+	  ,
+      mAVSyncTimeUs(-1)
+#endif
+{
     CHECK_EQ(mClient.connect(), (status_t)OK);
 
     DataSource::RegisterDefaultSniffers();
@@ -1617,6 +1622,17 @@ void AwesomePlayer::setVideoSource(sp<MediaSource> source) {
     mVideoTrack = source;
 }
 
+#ifdef MTK_HARDWARE
+void AwesomePlayer::mtk_omx_get_current_time(int64_t* pReal_time) {
+    if((mFlags & FIRST_FRAME) || mSeeking == SEEK) {
+            *pReal_time = -1;
+    } else {
+            *pReal_time = mAVSyncTimeUs;
+    }
+
+}
+#endif
+
 status_t AwesomePlayer::initVideoDecoder(uint32_t flags) {
     ATRACE_CALL();
 
@@ -1936,6 +1952,9 @@ void AwesomePlayer::onVideoEvent() {
         int64_t nowUs = estimateRealTimeUs(ts, systemTimeUs) - mTimeSourceDeltaUs;
 
         int64_t latenessUs = nowUs - timeUs;
+#ifdef MTK_HARDWARE
+        mAVSyncTimeUs = nowUs;
+#endif
 
         ATRACE_INT("Video Lateness (ms)", latenessUs / 1E3);
 
diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index ca310bc2d7a..d2a6896dd58 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -1797,10 +1797,19 @@ status_t OMXCodec::allocateOutputBuffersFromNativeWindow() {
 #endif
     }
 
+#ifdef MTK_HARDWARE
+    usage |= (GRALLOC_USAGE_SW_WRITE_OFTEN | GRALLOC_USAGE_SW_READ_OFTEN);
+#endif
+
     err = setNativeWindowSizeFormatAndUsage(
             mNativeWindow.get(),
+#ifdef MTK_HARDWARE
+            def.format.video.nStride,
+            def.format.video.nSliceHeight,
+#else
             def.format.video.nFrameWidth,
             def.format.video.nFrameHeight,
+#endif
             def.format.video.eColorFormat,
             rotationDegrees,
             usage | GRALLOC_USAGE_HW_TEXTURE | GRALLOC_USAGE_EXTERNAL_DISP);
diff --git a/media/libstagefright/include/AwesomePlayer.h b/media/libstagefright/include/AwesomePlayer.h
index 1a8e6c8e315..dddd49a3d69 100644
--- a/media/libstagefright/include/AwesomePlayer.h
+++ b/media/libstagefright/include/AwesomePlayer.h
@@ -107,6 +107,11 @@ struct AwesomePlayer {
     void postAudioEOS(int64_t delayUs = 0ll);
     void postAudioSeekComplete();
     void postAudioTearDown();
+
+#ifdef MTK_HARDWARE
+    void mtk_omx_get_current_time(int64_t* pReal_time);
+#endif
+
     status_t dump(int fd, const Vector<String16> &args) const;
 
     status_t suspend();
@@ -375,6 +380,9 @@ struct AwesomePlayer {
 
     AwesomePlayer(const AwesomePlayer &);
     AwesomePlayer &operator=(const AwesomePlayer &);
+#ifdef MTK_HARDWARE
+    int64_t mAVSyncTimeUs;
+#endif
 };
 
 }  // namespace android
